nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-nar-utils/src/main/java/org/apache/nifi/nar/StandardExtensionDiscoveringManager.java

```diff
@@ -383,42 +388,69 @@ public InstanceClassLoader createInstanceClassLoader(final String classType, fin
+                                || ancestorNarBundle.getBundleDetails().getCoordinate().getId().equals(NarClassLoaders.JETTY_NAR_ID)) {
+                                break;
+                            }
 
-            if (requiresInstanceClassLoading.cloneAncestorResources()) {
-                final ConfigurableComponent component = getTempComponent(classType, bundle.getBundleDetails().getCoordinate());
-                final Set<BundleCoordinate> reachableApiBundles = findReachableApiBundles(component);
+                            final NarClassLoader ancestorNarClassLoader = (NarClassLoader) ancestorClassLoader;
 
-                while (ancestorClassLoader instanceof NarClassLoader) {
-                    final Bundle ancestorNarBundle = classLoaderBundleLookup.get(ancestorClassLoader);
+                            narNativeLibDirs.add(ancestorNarClassLoader.getNARNativeLibDir());
+                            Collections.addAll(instanceUrls, ancestorNarClassLoader.getURLs());
 
-                    // stop including ancestor resources when we reach one of the APIs, or when we hit the Jetty NAR
-                    if (ancestorNarBundle == null || reachableApiBundles.contains(ancestorNarBundle.getBundleDetails().getCoordinate())
-                            || ancestorNarBundle.getBundleDetails().getCoordinate().getId().equals(NarClassLoaders.JETTY_NAR_ID)) {
-                        break;
+                            ancestorClassLoader = ancestorNarClassLoader.getParent();
+                        }
                     }
+                }
 
-                    final NarClassLoader ancestorNarClassLoader = (NarClassLoader) ancestorClassLoader;
-
-                    narNativeLibDirs.add(ancestorNarClassLoader.getNARNativeLibDir());
-                    Collections.addAll(instanceUrls, ancestorNarClassLoader.getURLs());
+                // register our new InstanceClassLoader as the shared base classloader.
+                if (baseClassLoaderKey != null && !resolvedSharedClassLoader) {
+                    // Created a shared class loader that is everything we need except for the additional URLs, as the additional URLs are instance-specific.
+                    final ClassLoader sharedClassLoader = new SharedInstanceClassLoader(instanceIdentifier, classType, instanceUrls, Collections.emptySet(), narNativeLibDirs, ancestorClassLoader);
+                    instanceClassLoader = new InstanceClassLoader(instanceIdentifier, classType, Collections.emptySet(), additionalUrls, Collections.emptySet(), sharedClassLoader);
 
-                    ancestorClassLoader = ancestorNarClassLoader.getParent();
+                    logger.debug("Creating InstanceClassLoader for type {} using newly created shared Base ClassLoader {} for component {}", type, instanceClassLoader, instanceIdentifier);
+                    sharedBaseClassloaders.putIfAbsent(baseClassLoaderKey, sharedClassLoader);
```

Comment:
Can we add logging that prints the resources of the shared class loader here? Something like...
```
if (logger.isTraceEnabled()) {
    for (URL url : ((SharedInstanceClassLoader)sharedClassLoader).getURLs()) {
      logger.trace("Shared Base ClassLoader URL resource: {}", new Object[] {url.toExternalForm()});
    }
}
```

We have similar logging at the end of this method for the InstanceClassLoader, which previously was helpful to debug and see all the resources that got combined, but now all of the resources can end up in the SharedInstanceClassLoader.